README.md
# pkk.js

[![Build Status](https://img.shields.io/circleci/project/github/stepankuzmin/pkk.js.svg?style=popout)](https://circleci.com/gh/stepankuzmin/pkk.js)
![npm](https://img.shields.io/npm/v/pkk)
![npm](https://img.shields.io/npm/dt/pkk)
![npm bundle size](https://img.shields.io/bundlephobia/minzip/pkk)

JavaScript Client for Public Cadastral Map ([pkk5.rosreestr.ru](https://pkk5.rosreestr.ru))

## Install

```shell
npm install pkk
```

## Usage

```js
import PKK, { FEATURE_TYPES } from 'pkk';

const pkk = new PKK();

pkk
  .queryFeatures(
    FEATURE_TYPES.LAND_PLOT,
    { lng: 37.629, lat: 55.7252 },
    { tolerance: 100, limit: 12 }
  )
  .then((features) => {
    const feature = features[0];
    const featureInfo = pkk.getFeatureInfo(FEATURE_TYPES.LAND_PLOT, feature.attrs.id);
    return featureInfo;
  })
  .then((featureInfo) => {
    console.log(pkk.formatFeatureInfo(featureInfo));
  });
```

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [PKK](#pkk)
    -   [Parameters](#parameters)
    -   [queryFeatures](#queryfeatures)
        -   [Parameters](#parameters-1)
        -   [Examples](#examples)
    -   [getFeatureInfo](#getfeatureinfo)
        -   [Parameters](#parameters-2)
        -   [Examples](#examples-1)
    -   [formatFeatureInfo](#formatfeatureinfo)
        -   [Parameters](#parameters-3)
        -   [Examples](#examples-2)
-   [Feature](#feature)
    -   [Properties](#properties)
-   [FeatureAttributes](#featureattributes)
    -   [Properties](#properties-1)
-   [FEATURE_TYPES](#feature_types)
    -   [Properties](#properties-2)

### PKK

PKK API Client

#### Parameters

-   `config` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** client config (optional, default `defaultConfig`)
    -   `config.baseURL` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** PKK base URL (optional, default `'https://pkk5.rosreestr.ru/'`)
    -   `config.featuresURL` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** PKK features API URL (optional, default `'/api/features/'`)
    -   `config.referer` **([Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean) \| [String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String))** referer header (optional, default `false`)
    -   `config.lang` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** i18n language (optional, default `'ru'`)

#### queryFeatures

Query PKK features near point

##### Parameters

-   `typeId` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** feature type id
-   `lnglat` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** longitude and latitude
    -   `lnglat.lng` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** longitude
    -   `lnglat.lat` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** latitude
-   `options` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** query options (optional, default `{tolerance:100,limit:12}`)
    -   `options.tolerance` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** query tolerance
    -   `options.limit` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** features query limit

##### Examples

```javascript
import PKK, { FEATURE_TYPES } from 'pkk';

const pkk = new PKK();

pkk.queryFeatures(
  FEATURE_TYPES.LAND_LOT,
  { lng: 37.629, lat: 55.7252 },
  { tolerance: 100, limit: 12 }
).then((features) => {
  console.log(features);
});
```

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[Feature](#feature)>>** response

#### getFeatureInfo

Query PKK feature info

##### Parameters

-   `typeId` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** feature type id
-   `featureId` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** feature id

##### Examples

```javascript
import PKK, { FEATURE_TYPES } from 'pkk';

const pkk = new PKK();

pkk.getFeatureInfo(FEATURE_TYPES.LAND_LOT, '77:1:1013:4985')
  .then((featureInfo) => {
    console.log(featureInfo);
  });
```

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[Feature](#feature)>** response

#### formatFeatureInfo

Formats feature info with i18n labels

##### Parameters

-   `featureInfo` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** feature info (optional, default `{}`)

##### Examples

```javascript
import PKK, { FEATURE_TYPES } from 'pkk';

const pkk = new PKK();

pkk.getFeatureInfo(FEATURE_TYPES.LAND_LOT, '77:1:1013:4985')
  .then((featureInfo) => {
    console.log(pkk.formatFeatureInfo(featureInfo));
  });
```

Returns **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** featureInfo translated feature info

### Feature

Объект

Type: [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

-   `type` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Тип объекта
-   `attrs` **[FeatureAttributes](#featureattributes)** Cписок атрибутов
-   `center` **{x: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), y: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)}** Центр объекта
-   `extent` **{xmax: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), xmin: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), ymax: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), ymin: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)}** Экстент объекта

### FeatureAttributes

Cписок атрибутов

Type: [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

-   `id` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Идентификатор
-   `cn` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Кадастровый номер
-   `statecd` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Статус
-   `address` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Адрес
-   `kladr` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Код КЛАДР
-   `fp` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Форма собственности
-   `adate` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Дата выгрузки сведений из ГКН
-   `anno_text` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Надпись на карте
-   `area_type` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Тип площади
-   `area_unit` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Единицы измерения площади
-   `area_value` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Декларированная площадь
-   `cad_cost` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Кадастровая стоимость
-   `cad_eng_data` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Кадастровый инженер
-   `cad_record_date` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Дата изменения сведений в ГКН
-   `cad_unit` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Единицы измерения стоимости
-   `category_type` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Категория земель
-   `date_cost` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Дата внесения кадастровой стоимости
-   `date_create` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Дата постановки на учет
-   `kvartal_cn` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Кадастровый номер квартала
-   `kvartal` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Квартал
-   `okrug_cn` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Кадастровый номер округа
-   `okrug` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Округ
-   `pubdate` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Дата опубликования на ПКК
-   `rayon_cn` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Кадастровый номер района
-   `rayon` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Район
-   `reg` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Номер субъекта
-   `rifr_cnt` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Контактное лицо
-   `rifr_dep` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Орган власти
-   `rifr` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Свободен от прав третьих лиц
-   `rights_reg` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Зарегистрированы права (да/нет)
-   `sale_cnt` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Контактное лицо
-   `sale_date` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Дата проведения торгов
-   `sale_dep` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Орган власти
-   `sale_price` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Начальная цена
-   `sale` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Принято решение о проведении торгов
-   `util_by_doc` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Разрешенное использование по документу
-   `util_code` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Разрешенное использование
-   `cc_date_approval` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Дата утверждения категории земель

### FEATURE_TYPES

Типы объектов

Type: [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

-   `LAND_LOT` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Земельные участки (1)
-   `LAND_QUARTER` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Кварталы (2)
-   `LAND_AREA` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Районы (3)
-   `LAND_DISTRICT` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Округи (4)
-   `CCO` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** ОКСы (5)
-   `TERRITORIAL_AREA` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Территориальные зоны (6)
-   `BOUNDARY` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Границы (7)
-   `BOUNDARY_REFERENCE_NETWORK` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** ОМС (9)
-   `USE_RESTRICTED_ZONE` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** ЗОУИТы (10)
-   `FORESTRY` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Лесничества и лесопарки (12)
-   `BUILDING_LINE` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Красные линии (13)
-   `LAND_LOT_LAYOUT` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Схемы расположения ЗУ (15)
-   `FREE_ECONOMIC_ZONE` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Свободные экономические зоны (16)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.1](https://github.com/stepankuzmin/pkk.js/compare/v0.2.0...v0.2.1) (2019-09-15)


### Features

* add AREA_TYPES, STATE_TYPES, FP_TYPES, CCO_TYPES, UNITS, and UTIL_TYPES constants ([75853ce](https://github.com/stepankuzmin/pkk.js/commit/75853ce))
* add docgen ([3b00985](https://github.com/stepankuzmin/pkk.js/commit/3b00985))
* add formatFeatureInfo method ([2654fbc](https://github.com/stepankuzmin/pkk.js/commit/2654fbc))

## [0.2.0](https://github.com/stepankuzmin/pkk.js/compare/v0.1.0...v0.2.0) (2019-09-14)


### ⚠ BREAKING CHANGES

* queryFeatures and getFeatureInfo requires feature type id argument

### Bug Fixes

* move referer header to config ([84de152](https://github.com/stepankuzmin/pkk.js/commit/84de152)), closes [#1](https://github.com/stepankuzmin/pkk.js/issues/1)


### Features

* add FEATURE_TYPES constants ([2e79311](https://github.com/stepankuzmin/pkk.js/commit/2e79311))

## 0.1.0 (2019-09-08)


### Features

* add getFeatureInfo ([973767a](https://github.com/stepankuzmin/pkk.js/commit/973767a))
* initial commit ([aaa54a3](https://github.com/stepankuzmin/pkk.js/commit/aaa54a3))


tests/index.test.js
import PKK, { FEATURE_TYPES } from '../src';

test('Features query', async () => {
  expect.assertions(1);

  const pkk = new PKK({ referer: true });

  const response = await pkk.queryFeatures(
    FEATURE_TYPES.LAND_LOT,
    { lng: 37.629, lat: 55.7252 },
    { tolerance: 100, limit: 12 },
  );

  expect(response).toEqual(
    expect.arrayContaining([
      expect.objectContaining({
        attrs: expect.objectContaining({
          address: expect.any(String),
          cn: expect.any(String),
          id: expect.any(String),
        }),
        center: expect.objectContaining({
          x: expect.any(Number),
          y: expect.any(Number),
        }),
        extent: expect.objectContaining({
          xmax: expect.any(Number),
          xmin: expect.any(Number),
          ymax: expect.any(Number),
          ymin: expect.any(Number),
        }),
        sort: expect.any(Number),
        type: expect.any(Number),
      }),
    ]),
  );
});

test('Feature info query', async () => {
  expect.assertions(1);

  const pkk = new PKK({ referer: true });
  const response = await pkk.getFeatureInfo(FEATURE_TYPES.LAND_LOT, '77:1:1013:4985');

  expect(response).toMatchObject({
    attrs: expect.objectContaining({
      adate: expect.any(String),
      address: expect.any(String),
      anno_text: expect.any(String),
      area_type: expect.any(String),
      area_unit: expect.any(String),
      area_value: expect.any(Number),
      cad_cost: expect.any(Number),
      cad_eng_data: expect.objectContaining({}),
      cad_record_date: expect.any(String),
      cad_unit: expect.any(String),
      category_type: expect.any(String),
      cn: expect.any(String),
      date_cost: expect.any(String),
      date_create: expect.any(String),
      id: expect.any(String),
      kvartal: expect.any(String),
      kvartal_cn: expect.any(String),
      okrug: expect.any(String),
      okrug_cn: expect.any(String),
      pubdate: expect.any(String),
      rayon: expect.any(String),
      rayon_cn: expect.any(String),
      reg: expect.any(Number),
      rights_reg: expect.any(Number),
      statecd: expect.any(String),
      util_by_doc: expect.any(String),
      util_code: expect.any(String),
    }),
    center: expect.objectContaining({
      x: expect.any(Number),
      y: expect.any(Number),
    }),
    extent: expect.objectContaining({
      xmax: expect.any(Number),
      xmin: expect.any(Number),
      ymax: expect.any(Number),
      ymin: expect.any(Number),
    }),
    type: expect.any(Number),
    stat: expect.objectContaining({}),
  });
});

test('Format Feature info', async () => {
  expect.assertions(1);

  const pkk = new PKK({ referer: true });
  const response = await pkk.getFeatureInfo(FEATURE_TYPES.LAND_LOT, '77:1:1013:4985');

  const featureInfo = pkk.formatFeatureInfo(response);

  const expectLabeledValue = expect.objectContaining({
    label: expect.any(String),
    value: expect.anything(),
  });

  expect(featureInfo).toMatchObject({
    attrs: expect.objectContaining({
      adate: expectLabeledValue,
      address: expectLabeledValue,
      anno_text: expectLabeledValue,
      area_type: expectLabeledValue,
      area_value: expectLabeledValue,
      cad_cost: expectLabeledValue,
      cad_eng_data: expectLabeledValue,
      cad_record_date: expectLabeledValue,
      category_type: expectLabeledValue,
      cn: expectLabeledValue,
      date_cost: expectLabeledValue,
      date_create: expectLabeledValue,
      id: expectLabeledValue,
      kvartal: expectLabeledValue,
      kvartal_cn: expectLabeledValue,
      okrug: expectLabeledValue,
      okrug_cn: expectLabeledValue,
      pubdate: expectLabeledValue,
      rayon: expectLabeledValue,
      rayon_cn: expectLabeledValue,
      reg: expectLabeledValue,
      rights_reg: expectLabeledValue,
      statecd: expectLabeledValue,
      util_by_doc: expectLabeledValue,
      util_code: expectLabeledValue,
    }),
    center: expect.objectContaining({
      x: expect.any(Number),
      y: expect.any(Number),
    }),
    extent: expect.objectContaining({
      xmax: expect.any(Number),
      xmin: expect.any(Number),
      ymax: expect.any(Number),
      ymin: expect.any(Number),
    }),
    type: expect.any(Number),
    stat: expect.objectContaining({}),
  });
});


src/docgen.js
import fs from 'fs';
import path from 'path';
import Polyglot from 'node-polyglot';
import translations from './translations';
import { FEATURE_TYPES } from './constants';

const phrases = translations.ru;
const polyglot = new Polyglot({ phrases });

const feature = (output) => {
  output.write(`/**
 * Объект
 * @typedef {Object} Feature
 * @property {number} type ${polyglot.t('feature.type')}
 * @property {FeatureAttributes} attrs ${polyglot.t('feature.attrs')}
 * @property {{x: number, y: number}} center ${polyglot.t('feature.center')}
 * @property {{xmax: number, xmin: number, ymax: number, ymin: number}} extent ${polyglot.t(
    'feature.extent',
  )}
 */\n\n`);
};

const featureAttributes = (output) => {
  output.write(`/**
 * Cписок атрибутов
 * @typedef {Object} FeatureAttributes\n`);

  Object.keys(phrases.attrs[FEATURE_TYPES.LAND_LOT]).forEach((attr) => {
    const description = polyglot.t(`attrs.${FEATURE_TYPES.LAND_LOT}.${attr}`);
    const param = ` * @property {String} ${attr} ${description}\n`;
    output.write(param);
  });

  output.write(' */\n\n');
};

const featureTypes = (output) => {
  output.write(`/**
 * Типы объектов
 * @typedef {Object} FEATURE_TYPES\n`);

  Object.keys(FEATURE_TYPES).forEach((FEATURE_TYPE) => {
    const typeId = FEATURE_TYPES[FEATURE_TYPE];
    const description = polyglot.t(`TYPE_IDS.${typeId}`);
    const param = ` * @property {number} ${FEATURE_TYPE} ${description} (${typeId})\n`;
    output.write(param);
  });

  output.write(' */\n');
};

const outputPath = path.join(__dirname, './definitions.js');
const output = fs.createWriteStream(outputPath);
output.write('/* This file is autogenerated, do not edit */\n\n');

feature(output);
featureAttributes(output);
featureTypes(output);

output.close();


src/definitions.js
/* This file is autogenerated, do not edit */

/**
 * Объект
 * @typedef {Object} Feature
 * @property {number} type Тип объекта
 * @property {FeatureAttributes} attrs Cписок атрибутов
 * @property {{x: number, y: number}} center Центр объекта
 * @property {{xmax: number, xmin: number, ymax: number, ymin: number}} extent Экстент объекта
 */

/**
 * Cписок атрибутов
 * @typedef {Object} FeatureAttributes
 * @property {String} id Идентификатор
 * @property {String} cn Кадастровый номер
 * @property {String} statecd Статус
 * @property {String} address Адрес
 * @property {String} kladr Код КЛАДР
 * @property {String} fp Форма собственности
 * @property {String} adate Дата выгрузки сведений из ГКН
 * @property {String} anno_text Надпись на карте
 * @property {String} area_type Тип площади
 * @property {String} area_unit Единицы измерения площади
 * @property {String} area_value Декларированная площадь
 * @property {String} cad_cost Кадастровая стоимость
 * @property {String} cad_eng_data Кадастровый инженер
 * @property {String} cad_record_date Дата изменения сведений в ГКН
 * @property {String} cad_unit Единицы измерения стоимости
 * @property {String} category_type Категория земель
 * @property {String} date_cost Дата внесения кадастровой стоимости
 * @property {String} date_create Дата постановки на учет
 * @property {String} kvartal_cn Кадастровый номер квартала
 * @property {String} kvartal Квартал
 * @property {String} okrug_cn Кадастровый номер округа
 * @property {String} okrug Округ
 * @property {String} pubdate Дата опубликования на ПКК
 * @property {String} rayon_cn Кадастровый номер района
 * @property {String} rayon Район
 * @property {String} reg Номер субъекта
 * @property {String} rifr_cnt Контактное лицо
 * @property {String} rifr_dep Орган власти
 * @property {String} rifr Свободен от прав третьих лиц
 * @property {String} rights_reg Зарегистрированы права (да/нет)
 * @property {String} sale_cnt Контактное лицо
 * @property {String} sale_date Дата проведения торгов
 * @property {String} sale_dep Орган власти
 * @property {String} sale_price Начальная цена
 * @property {String} sale Принято решение о проведении торгов
 * @property {String} util_by_doc Разрешенное использование по документу
 * @property {String} util_code Разрешенное использование
 * @property {String} cc_date_approval Дата утверждения категории земель
 */

/**
 * Типы объектов
 * @typedef {Object} FEATURE_TYPES
 * @property {number} LAND_LOT Земельные участки (1)
 * @property {number} LAND_QUARTER Кварталы (2)
 * @property {number} LAND_AREA Районы (3)
 * @property {number} LAND_DISTRICT Округи (4)
 * @property {number} CCO ОКСы (5)
 * @property {number} TERRITORIAL_AREA Территориальные зоны (6)
 * @property {number} BOUNDARY Границы (7)
 * @property {number} BOUNDARY_REFERENCE_NETWORK ОМС (9)
 * @property {number} USE_RESTRICTED_ZONE ЗОУИТы (10)
 * @property {number} FORESTRY Лесничества и лесопарки (12)
 * @property {number} BUILDING_LINE Красные линии (13)
 * @property {number} LAND_LOT_LAYOUT Схемы расположения ЗУ (15)
 * @property {number} FREE_ECONOMIC_ZONE Свободные экономические зоны (16)
 */


src/index.js
import axios from 'axios';
import Polyglot from 'node-polyglot';
import translations from './translations';
import './definitions';

const defaultConfig = {
  baseURL: 'https://pkk5.rosreestr.ru/',
  featuresURL: '/api/features',
  referer: false,
  lang: 'ru',
};

/**
 * PKK API Client
 * @param {Object} config client config
 * @param {String} [config.baseURL='https://pkk5.rosreestr.ru/'] PKK base URL
 * @param {String} [config.featuresURL='/api/features/'] PKK features API URL
 * @param {Boolean|String} [config.referer=false] referer header
 * @param {String} [config.lang='ru'] i18n language
 */
class PKK {
  constructor(config = defaultConfig) {
    this.config = { ...defaultConfig, ...config };

    const phrases = translations[this.config.lang];
    this.polyglot = new Polyglot({ phrases });

    const headers = {};
    if (this.config.referer) {
      headers.referer = typeof this.config.referer === 'string' ? this.config.referer : this.config.baseURL;
    }

    this.axios = axios.create({ baseURL: this.config.baseURL, headers });
  }

  /**
   * Query PKK features near point
   * @param {Number} typeId feature type id
   * @param {Object} lnglat longitude and latitude
   * @param {Number} lnglat.lng longitude
   * @param {Number} lnglat.lat latitude
   * @param {Object} options query options
   * @param {Number} options.tolerance query tolerance
   * @param {Number} options.limit features query limit
   * @returns {Promise<Array<Feature>>} response
   * @example
   * import PKK, { FEATURE_TYPES } from 'pkk';
   *
   * const pkk = new PKK();
   *
   * pkk.queryFeatures(
   *   FEATURE_TYPES.LAND_LOT,
   *   { lng: 37.629, lat: 55.7252 },
   *   { tolerance: 100, limit: 12 }
   * ).then((features) => {
   *   console.log(features);
   * });
   */
  queryFeatures = (typeId, { lng, lat }, options = { tolerance: 100, limit: 12 }) => {
    const params = {
      text: `${lat},${lng}`,
      limit: options.limit,
      tolerance: options.tolerance,
    };

    const featuresURL = `${this.config.featuresURL}/${typeId}`;
    return this.axios.get(featuresURL, { params }).then((response) => {
      const { data } = response;
      if (data.status !== 200) {
        throw new Error(data.note);
      }

      return data.features;
    });
  };

  /**
   * Query PKK feature info
   * @param {Number} typeId feature type id
   * @param {String} featureId feature id
   * @returns {Promise<Feature>} response
   * @example
   * import PKK, { FEATURE_TYPES } from 'pkk';
   *
   * const pkk = new PKK();
   *
   * pkk.getFeatureInfo(FEATURE_TYPES.LAND_LOT, '77:1:1013:4985')
   *   .then((featureInfo) => {
   *     console.log(featureInfo);
   *   });
   */
  getFeatureInfo = (typeId, featureId) => {
    const featureURL = `${this.config.featuresURL}/${typeId}/${featureId}`;
    return this.axios.get(featureURL).then((response) => {
      const { data } = response;
      if (data.status !== 200) {
        throw new Error(data.note);
      }

      return data.feature;
    });
  };

  /**
   * Formats feature info with i18n labels
   * @param {Object} featureInfo feature info
   * @returns {Object} featureInfo translated feature info
   * @example
   * import PKK, { FEATURE_TYPES } from 'pkk';
   *
   * const pkk = new PKK();
   *
   * pkk.getFeatureInfo(FEATURE_TYPES.LAND_LOT, '77:1:1013:4985')
   *   .then((featureInfo) => {
   *     console.log(pkk.formatFeatureInfo(featureInfo));
   *   });
   */
  formatFeatureInfo = (featureInfo = {}) => {
    /* eslint-disable camelcase */
    const { type } = featureInfo;

    const {
      area_type,
      area_unit,
      area_value,
      cad_cost,
      cad_eng_data,
      cad_unit,
      category_type,
      fp,
      statecd,
      util_code,
      ...attrs
    } = featureInfo.attrs;

    const translatedAttrs = {
      area_value: this.formatAreaValue(featureInfo),
      cad_cost: this.formatCadastralCost(featureInfo),
      cad_eng_data: this.formatCadastralEngineer(featureInfo),
    };

    if (area_type) {
      translatedAttrs.area_type = {
        label: this.polyglot.t(`attrs.${type}.area_type`),
        value: this.polyglot.t(`AREA_TYPES.${area_type}`),
      };
    }

    if (category_type) {
      translatedAttrs.category_type = {
        label: this.polyglot.t(`attrs.${type}.category_type`),
        value: this.polyglot.t(`CATEGORY_TYPES.${category_type}`),
      };
    }

    if (statecd) {
      translatedAttrs.statecd = {
        label: this.polyglot.t(`attrs.${type}.statecd`),
        value: this.polyglot.t(`STATE_TYPES.${statecd}`),
      };
    }

    if (fp) {
      translatedAttrs.fp = {
        label: this.polyglot.t(`attrs.${type}.fp`),
        value: this.polyglot.t(`FP_TYPES.${fp}`),
      };
    }

    if (util_code) {
      translatedAttrs.util_code = {
        label: this.polyglot.t(`attrs.${type}.util_code`),
        value: this.polyglot.t(`UTIL_TYPES.${util_code}`),
      };
    }

    Object.keys(attrs).forEach((attr) => {
      translatedAttrs[attr] = {
        label: this.polyglot.t(`attrs.${type}.${attr}`),
        value: attrs[attr],
      };
    });

    return { ...featureInfo, attrs: translatedAttrs };
    /* eslint-enable camelcase */
  };

  formatAreaValue = (featureInfo) => {
    const {
      type,
      attrs: { area_value: areaValue, area_unit: unitId },
    } = featureInfo;

    const label = this.polyglot.t(`attrs.${type}.area_value`);

    if (!areaValue || !unitId) {
      return { label, value: null };
    }

    const units = this.polyglot.t(`UNITS.${unitId}`);
    const value = `${areaValue} ${units}`;

    return { label, value };
  };

  formatCadastralCost = (featureInfo) => {
    const {
      type,
      attrs: { cad_cost: cost, cad_unit: unitId },
    } = featureInfo;

    const label = this.polyglot.t(`attrs.${type}.cad_cost`);

    if (!cost || !unitId) {
      return { label, value: null };
    }

    const units = this.polyglot.t(`UNITS.${unitId}`);
    const value = `${cost} ${units}`;

    return { label, value };
  };

  formatCadastralEngineer = ({ type, attrs: { cad_eng_data: data } }) => {
    const label = this.polyglot.t(`attrs.${type}.cad_eng_data`);

    if (!data) {
      return { label, value: null };
    }

    const value = data.rc_type === 0
      ? `${data.ci_surname} ${data.ci_first} ${data.ci_patronymic}`
      : data.co_name;

    return { label, value };
  };
}

export default PKK;
export * from './constants';
